### ISSUE TITLE
math test failures due to LLVM optimizer not respecting "nobuiltin" attribute (fixed upstream)

### ISSUE THREAD

**Issue by tiehuis** [zig-lang/zig#393]
See the `math` branch.

Some failures when running in release mode that need to be resolved before we can merge. The exact test cases that fail under release-fast are:

```
pow32(37.45, 3.3)
exp2f(0.2)
exp2f(0.8923)
exp2f(37.45)
exp2d(0.2)
exp2d(0.8923)
sin32(37.45)
sin32(89.123)
cos32(89.123)
tan32(1.5)
tan32(37.45)
tan32(89.123)
```

release-safe is also okay once commenting the following tests, so is likely the same.

----

```
$ (math) zig test --release-safe pow.zig 
Test 1/132 pow...OK
Test 2/132 pow32...reached unreachable code
???:?:?: 0x65726e7520646568 in ??? (???)
    ???


Tests failed. Use the following command to reproduce the failure:
./test
$ (math) zig test --release-fast pow.zig
Test 1/132 pow...OK
Test 2/132 pow32...reached unreachable code
/usr/lib/zig/std/special/compiler_rt.zig:15:24: 0x0000000000000003 in ??? (compiler_rt)
    return __udivmoddi4(a, b, null);
                       ^
???:?:?: 0x0000003436776f70 in ??? (???)
    ???


Tests failed. Use the following command to reproduce the failure:
./test
```

**Comment by andrewrk** [comment]
One thing to note here is that our stack unwinding code is a bit naive and does not take advantage of the call frame information provided in DWARF. In debug mode it works fine, but in release mode we have pretty useless stack traces - but we could improve the stack trace code and make them useful.

Regardless we can still use gdb to debug in release mode.

One of the tricks I like to do when debugging this kind of code, is have a C version and a Zig version side by side and step through them together in a debugger. Since the Zig code is ported from C, we can narrow down the problem to where they diverge.

**Comment by andrewrk** [comment]
I'm going to go ahead and schedule this for 0.1.0 since you already did most of the work.

**Comment by andrewrk** [comment]
here's a clue...

I changed the pow32 test so that it was only these 2:
```
test "pow32" {
    const epsilon = 0.000001;

    // assert(math.approxEq(f32, pow32(0.0, 3.3), 0.0, epsilon)); // TODO: Handle div zero
    //assert(math.approxEq(f32, pow32(0.8923, 3.3), 0.686572, epsilon));
    //assert(math.approxEq(f32, pow32(0.2, 3.3), 0.004936, epsilon));
    assert(math.approxEq(f32, pow32(1.5, 3.3), 3.811546, epsilon));
    assert(math.approxEq(f32, pow32(37.45, 3.3), 155736.703125, epsilon));
    //assert(math.approxEq(f32, pow32(89.123, 3.3), 2722489.5, epsilon));
}
```

And then use this command to only run these tests in isolation:

```
[nix-shell:~/dev/zig/build]$ ./zig test ../std/math/pow.zig --release-fast --test-filter pow32
Test 1/1 pow32...reached unreachable code

Tests failed. Use the following command to reproduce the failure:
./test
```

But if I comment out either one of the two tests, so that the other test runs in isolation, then it passes. It is almost as if there is some global state not being reset.

**Comment by andrewrk** [comment]
If I run the same test multiple times, but only that test and nothing else, it works fine:

```
    assert(math.approxEq(f32, pow32(37.45, 3.3), 155736.703125, epsilon));
    assert(math.approxEq(f32, pow32(37.45, 3.3), 155736.703125, epsilon));
    assert(math.approxEq(f32, pow32(37.45, 3.3), 155736.703125, epsilon));
    assert(math.approxEq(f32, pow32(37.45, 3.3), 155736.703125, epsilon));
```
```
Test 1/1 math.pow32...OK
```

**Comment by andrewrk** [comment]
If only 1 assert is present (even if it's the same one copy-pasted multiple times), LLVM evaluates the entire test at compile time and the test function has an empty body. When multiple different tests are present, then LLVM emits the pow32 function into the binary.

**Comment by andrewrk** [comment]
@tiehuis one thing I noticed, in pow32 we have:

```
    if (math.isInf(x)) {
        if (math.isNegativeInf(x)) {
```

but in pow64 we have:

```
    if (math.isInf(x)) {
        if (math.isInf(x)) {
```

It looks like the pow32 one is correct, but the pow64 one is the one with tests passing. Although I don't think we have tests to cover this particular thing, so perhaps it is a different issue.

**Comment by tiehuis** [comment]
Nice findings. I just checked and can reproduce everything as you say.

Yes, pow64 is incorrect. I'll need to go through and extend the tests to go through all special cases at some point.

**Comment by tiehuis** [comment]
I went through the tests again and cannot reproduce exp. Also, for the trigonometric functions, adding `@setFloatMode(this, @import("builtin").FloatMode.Strict)` will allow them all to pass okay. `pow` is a bit more complex and haven't yet got this to work.

One thing I actually forgot to note was that the failing functions were the ones I took from go instead of musl. So may have a closer look into this and any assumptions go makes.

**Comment by andrewrk** [comment]
pow can be implemented pretty easily to be generic with f32 and f64. I'll push what I have to the math branch. With this change the only test that fails is 

```
    assert(math.approxEq(f32, pow(f32, 89.123, 3.3), 2722489.5, epsilon));
```

in release mode.

**Comment by tiehuis** [comment]
I've commented that test out for the moment and switched over the other failing functions to use StrictMode where it fixes this.

This doesn't fix the underlying problem, but I'll start adding tests to cover edge cases for these functions and hopefully some more clues about this will appear in the progress.

**Comment by andrewrk** [comment]
This is baffling to me:

```
[nix-shell:~/dev/zig/build]$ ./zig test ../std/math/ceil.zig --release-fast  --test-filter ceil
Test 1/3 ceil...OK
Test 2/3 ceil32...OK
Test 3/3 ceil64...OK

[nix-shell:~/dev/zig/build]$ ./zig test ../std/math/ceil.zig --release-fast --test-name-prefix 'std-ReleaseFast-bare ' --test-filter ceil
Test 1/3 std-ReleaseFast-bare ceil...reached unreachable code
/home/andy/dev/zig/build/lib/zig/std/special/compiler_rt.zig:20:5: 0x0000000000000019 in ??? (compiler_rt)
    return *@ptrCast(&udwords, &x);
    ^
???:?:?: 0x2d74736146657361 in ??? (???)
    ???


Tests failed. Use the following command to reproduce the failure:
./test

```

I'll have a look tomorrow.

**Comment by tiehuis** [comment]
Here is a diff of the object dump of the above `ceil` tests which may help. The following is the only difference.

```
0000000000205090 <ceil>:
  205090:       8b 44 24 f8             mov    -0x8(%rsp),%eax
  205094:       89 44 24 f8             mov    %eax,-0x8(%rsp)
  205098:       8b 44 24 fc             mov    -0x4(%rsp),%eax
  20509c:       89 44 24 fc             mov    %eax,-0x4(%rsp)
  2050a0:       c3                      retq   
  2050a1:       66 66 66 66 66 66 2e    data16 data16 data16 data16 data16 nopw %cs:0x0(%rax,%rax,1)
  2050a8:       0f 1f 84 00 00 00 00 
  2050af:       00 
```

```
0000000000205090 <std-ReleaseFast-bare ceil>:
  205090:       50                      push   %rax
  205091:       8b 04 24                mov    (%rsp),%eax
  205094:       89 04 24                mov    %eax,(%rsp)
  205097:       8b 44 24 04             mov    0x4(%rsp),%eax
  20509b:       89 44 24 04             mov    %eax,0x4(%rsp)
  20509f:       48 8d 3d ba d7 ff ff    lea    -0x2846(%rip),%rdi        # 202860 <__unnamed_1>
  2050a6:       be 18 00 00 00          mov    $0x18,%esi
  2050ab:       e8 10 89 00 00          callq  20d9c0 <__zig_panic>
```

**Comment by andrewrk** [comment]
It looks like in both cases, the LLVM optimizer determined whether the test would pass at compile time. In the first one, it passed, and in the second one, it failed. This is perplexing to me given that the only difference is the name of the symbol. I'm looking into it.

**Comment by andrewrk** [comment]
I suspect this might actually be an LLVM bug. Here's why:

When I rename the tests in ceil.zig to:

```
test "math.ceil" 

test "math.ceil32" 

test "math.ceil64"
```

Now ceil test fails with `--release-fast` regardless of `--test-name-prefix`. What was happening there is that the test names actually become function names, and in the LLVM IR we were outputting the ceil test function before the ceil implementation function, so the actual ceil symbol name was mangled to something like `ceil.6`. When the ceil implementation had the mangled symbol name, it worked correctly.

However when we passed `--test-name-prefix` the test `ceil` was now called `test-ReleaseFast-std ceil` and so the actual ceil implementation function symbol name no longer had to be mangled and was called `ceil`.

So I suspect that LLVM is actually doing a string comparison on the function symbol name, and doing some different logic when it is called `ceil`.

If I change the name of `ceil` function to `ceil_asdf` (and modify the callers) then the tests pass.

**Comment by andrewrk** [comment]
I think we need to put this on all functions (from http://llvm.org/docs/LangRef.html#fnattrs):

```
nobuiltin
This indicates that the callee function at a call site is not recognized as a built-in function. LLVM will retain the original call and not replace it with equivalent code based on the semantics of the built-in function, unless the call site uses the builtin attribute. This is valid at call sites and on function declarations and definitions.
```

**Comment by andrewrk** [comment]
That did not solve the problem. Asking on the LLVM mailing list.

**Comment by andrewrk** [comment]
This is in fact an LLVM bug which just recently fixed: https://reviews.llvm.org/rL305132

As a temporary workaround, we can rename the math functions to be names that do not collide with the libc math names, and then expose them like this:

```
// TODO remove this workaround when we can depend on LLVM 4.0.1
pub const ceil = ceil_workaround;
```

I'm working on this right now and I'll push to the math branch.

Unrelated, a couple more things:
 * Why is _expo2 underscored?
 * Before merging we need to add the lines in CMakeLists.txt to make these math .zig files actually installed.

**Comment by andrewrk** [comment]
I'm removing `fmod` because `@rem`

**Comment by andrewrk** [comment]
OK I pushed a workaround for this and now the only function that is failing is pow.

**Comment by tiehuis** [comment]
Nice job tracking that down!

Regarding _expo2, that was to indicate it is internal. We don't really need that though so I'll rename it.

I'd also like to add the test special cases for all the functions where I can as well before any merge, so I'll modify the CMakeLists.txt in the process.

**Comment by tiehuis** [comment]
Made the mentioned changes above.

Just to note, after adding the special case tests the following also seemed to behave similar to pow on release build.
 - frexp
 - tanh
 - sinh
 - exp

**Comment by andrewrk** [comment]
For frexp, this test case fails when isolated:

```
    r = frexp64(math.inf_f64);
    assert(math.isPositiveInf(r.significand) and r.exponent == 0);
```

What's really strange to me is that if I do this:

```
    r = frexp64(math.inf_f64);
    assert(math.isPositiveInf(r.significand));
    assert(r.exponent == 0);
```

...then it passes. I'm going to have a look at the LLVM IR difference between these two things and see what's going on.

**Comment by andrewrk** [comment]
Found the bug in frexp - it does not set the exponent on the result, so the exponent is undefined.

```
    } else if (e == 0x7FF) {
        // frexp(nan) = (nan, 0)
        result.significand = x;
        return result;
```
```
    assert(r.exponent == 0); // r.exponent is undefined
```

the musl code doesn't set the exponent either in this case, so I wonder where you got this test case from?

**Comment by andrewrk** [comment]
for exp, does it have to do with this?

```
        // TODO: Determine how to do the following.
        // y *= 0x1.0p1023;
```

this is from scalbn which exp calls.

**Comment by andrewrk** [comment]
I also see these:

../std/math/tanh.zig:            t = 1.0; // TODO + 0 / x;
../std/math/expm1.zig:            y = y * 2.0; // TODO: * 0x1.0p1023;

**Comment by tiehuis** [comment]
The special test cases I've taken from go. These follow musl in some areas but differ in some others.

I've made changes to address these issues. Everything (besides pow still) compiles under release mode. Largely this is just adding explicit nan checks at the beginning of some blocks.

See https://github.com/zig-lang/zig/commit/14a324a0fa8d60dbf7d6948a992a09d8a65cbdc4 for some comments.

**Comment by tiehuis** [comment]
Ok, regarding pow, it seems like the error is related to this test:

```
assert(math.approxEq(f32, pow(f32, 37.45, 3.3), 155736.6875, epsilon));
```

In debug, it results in `~155736.703125`, which is as we expect, however in release mode we get a different result of `~155736.6875`. It seems like there is some too aggressive optimization taking place in pow or one of its called functions.

After going through the functions, setting `ln` to strict mode corrects the error.

I've pushed a fix for this, however I'd like to look into what release mode does on the llvm side. Seems like its applying some transformations that differ from safe mode, even though both should be using the same floating point modes?

**Comment by andrewrk** [comment]
> I'd like to look into what release mode does on the llvm side. Seems like it's applying some transformations that differ from safe mode, even though both should be using the same floating point modes?

One possibility is that safety checks prevent optimizations. For example, checking for division by zero might prevent floating point optimizations. So this is one way that release-fast and release-safe could be different.

> Also, remains to be seen why the assertion failure was resulting in the compiler_rt error.

This could be because we naively try to look at the base pointer register without consulting the DWARF stack unwinding info. If the base pointer register is used as a generic register, like it is in release mode, then this can result in seemingly random addresses dereferenced as stack addresses, which could be what caused the compiler rt stack trace thing. I created #398 for this.

**Comment by andrewrk** [comment]
Looks like tests are passing in the math branch! Ready to merge into master?

**Comment by tiehuis** [comment]
I'm ready when you are. I'll leave it to you to merge if that's okay; just made one small change to change frexp to the undefined result behavior.

**Comment by andrewrk** [comment]
Ok great I'll take it from here. Thanks again for all your hard work! Give me a few days as I'm out of town until Monday.

**Comment by andrewrk** [comment]
This will be fixed when we switch to depending on LLVM 5.0.0 which is scheduled for Zig 0.2.0.
